content/blog/2022-05-19-fast-brili.md

```diff
@@ -0,0 +1,141 @@
++++
+title = "Bril bytecode interpreter"
+[extra]
+bio = """
+	Charles Sherk is an junior studying Computer Science & Mathematics at
+	Cornell university, interested in compilers and programming language
+	implementation.
+	Susan Garry is a senior studying Computer Science at Cornell University,
+	interested in compilers and programming language design.
+
+"""
+[[extra.authors]]
+name = "Charles Sherk"
+link = "https://github.com/charles-rs/"
+[[extra.authors]]
+name = "Susan Garry"
+link = "https://github.com/susan-garry"
++++
+
+## Motivation
+Currently we have access to `brili`, which is written with an emphasis on
+readable code and extensiblity, and `brili-rs`, which is supposed to be faster,
+but `brili-rs` isn't actually that fast, and seems to be rather bloated for what
+it is. There were some points in this course where time was spent on "waiting
+for the Bril interpreter" that could have been avoided. So, we propose and
+implement an interpreter for Bril whose sole focus is speed.
+
+## Design overview
+
+### Language choice
+We chose C for an implementation language as it is known for its speed as well
+as simplicity, which means we won't inadvertently write code that is more
+expensive than necessary.
+
+### Big Red Bytecode
+In order to make our interpreter as efficient as possible, we have designed a
+very compressed representation of bril instructions: (almost) all of them are
+represented in 64 bits, with some taking multiple 64 bit words as needed. See
+the [documentation](https://github.coecis.cornell.edu/cs897/fast-bril/blob/master/doc/brb.pdf).
+
+In order to appease the C typechecker and not rely on undefined behavior, we use
+a union type to internally represent instructions, but most have the first 16
+bits as the opcode, so we can pull this out and then determine how to decode the
+rest of the instruction. In principle, we could also use `uint64_t` or
+`uint16_t[4]`, but using unions is the most user friendly, and lets the C
+compiler do more work to access the fields of structs, in case there are any
+cool tricks for that behind the scenes.
```

Comment:
The use of unions in the interpreter is subordinate to the design of the encoding itself. This paragraph gives a *hint* as to what the encoding looks like, but it doesn't quite spell it out. Can you show us an example or two of what the encoding actually looks like? Maybe even just an `add` instruction? Doing so would force you to describe how variables work in your bytecode, which would be a huge expansion in terms of specificity.

Again, imagine the audience as a generally compilers-curious person. They will want to know a teensy bit more about what this whole "bytecode representation" business really means, and examples could help a lot with that.

Reply:
think this is resolved by the documentation pdf?

Reply:
think this is resolved by the documentation pdf?